### ISSUE TITLE
Sound is laggy while game is running 60fps on iPad air

### ISSUE THREAD

**Issue by essiw** [WilliamLCobb/iNDS#86]
While running games on a iPad Air (version 1) on full speed (60fps) the sound has slight lag. This is really frustrating while playing games like pokemon (I tried platinum), where the sound sounds just a tiny bit slower than it should be. I guess it takes more resources to handle full sound but I would think that an iPad Air could handle full sound, as it is quite a quick device. So I would like to ask if it is possible to enhance the game a tiny bit so an iPad Air can play these games with normal sound. To me playing games without sound is no option. Also the synchronised sound option is not an option as well because it sounds horrible :|

I certainly hope there is a solution for this. It seems like it is really lagging only a tiny bit.

**Comment by essiw** [comment]
maybe the idea in issue 85 will solve this problem: https://github.com/WilliamLCobb/iNDS/issues/85
7. Option for loading game into memory (which should help some games with audio lags).

If this might solve issue and you are closing this issue because of that would it be possible to indicate what priority this fix has? :) Thanks

**Comment by DarknesGaming** [comment]
I'm having the same problem on my iPhone 6. There's probably going to be an iNDS 1.4.1 update to patch the whole lag thing. @WilliamLCobb will probably fix the audio lag thing, and the synchronized sound option. And maybe the hiccup thing because the emulation would stop for a tick and then keep going.
Also, if you load the game onto the RAM,  it causes high memory usage and may lead to problems. By default, DeSmuME loads the game onto an ISO (or something related to a disk image file), which won't cause high memory usage. Since there's no option on iNDS to choose whether to load a ROM onto an ISO or RAM, there's really no way to change that. In case you don't know, RAM is short for Random Access Memory.
The iPad Air is indeed quick, but it's trying to run an app with over 400 issues as smooth as possible.
tl;dr summary: I'm having the same problem on my iPhone, @WilliamLCobb will probably fix it. The end.
iOS Device: iPhone 6 (iPhone 7,2)
iOS version: 9.3
Jailbreak status: No
Download location: Compiled source code with Xcode
Current iNDS Version: 1.4.0.

**Comment by essiw** [comment]
I certainly hope he will :)

**Comment by DarknesGaming** [comment]
Wow. Looks like @WilliamLCobb has more bugs to fix. You seem to be really active with bug hunting.

**Comment by DarknesGaming** [comment]
Reasons why sound is laggy: I have no idea. Perhaps work on the sound part of the DeSmuME system?
I honestly have no idea. I'm trying to figure out if TinyXML should remain at the version it's on now or upgrade to TinyXML2. The developer of TinyXML admitted that it used memory inefficiently and did NOT perform "as well as desired for mobile devices." That's why he created TinyXML2, which is somehow lighter.

**Comment by essiw** [comment]
@DarknesGaming While I appreciate your effort in trying to help I am wondering what added value some (parts) of your comments have. I am not trying to be mean, I just want to be sure the problem gets to the developer and he does not have to spend more time than needed.

**Comment by DarknesGaming** [comment]
I understand what you mean @essiw. I'm extremely new to the DeSmuME code, so @WilliamLCobb can handle that.
Since we have the beta branch, I'll try to fix the problems, but because I'm new, there's a high chance that it may not work.

**Comment by essiw** [comment]
@WilliamLCobb Just wondering if this is seen as an issue, or just seen as slow performance. I am also wondering what priority this issue has. And what the priority is of the issue about the hiccups in the sound.

**Comment by pmp174** [comment]
@essiw did you sideload the app on debug mode or release?

**Comment by essiw** [comment]
@pmp174 I am unsure what you mean with sideload. I loaded the app (1.4.0) the normal way, as I have no developer skills (except knowledge of html and css). So I loaded it on my iPad through the app downloaded from the buildstore (builds.io).

To add to this the issue (except for the hiccups in the sound) seems not to happen with games that are not resource heavy (like Phoenix Wright Ace Attorney)

**Comment by DarknesGaming** [comment]
@pmp174 I think essiw loaded the app on release and not debug. If it's from BuildStore, it's probably on release.

**Comment by WilliamLCobb** [comment]
Functionality take precedence over things that aren't necessary. So after the microphone is finished I'll address this. I don't think fixing the sound will be too hard, it definitely is dependent on frame rate

**Comment by essiw** [comment]
@WilliamLCobb Great :D thanks for your reply. I cannot wait for this to be fixed :)

**Comment by DarknesGaming** [comment]
Well, this is a good thing, because the synchronized sound option cannot sound horrible anymore. The microphone will be significant though, and maybe there will be a 1.4.1 for that.
@essiw I should probably tell you this now: The older the ROM is, the more likely it's going to run slower for obvious reasons. If it's not obvious, then you need time to figure that out.

**Comment by DarknesGaming** [comment]
9:04 PM:Time to see if the sound is not lagging. Wish me luck. Someone needs to tell me what the (XX core) means.

**Comment by essiw** [comment]
Indeed the sound lag is still there, it might have been become a bit better but I am not sure about that.

**Comment by FrogKnight** [comment]
The sound is much better than before but still is not perfect.

**Comment by DarknesGaming** [comment]
That is very true. You can really tell the sound is lagging if you play Pokemon Black 2/White 2 and battle a trainer during a sandstorm.

**Comment by essiw** [comment]
For the reference, I am comparing the sound with Pokemon Diamond, even outside battles the sound stil lag. It seems to be improved a bit but not enough.

**Comment by DarknesGaming** [comment]
I tried playing Pokemon Diamond. It's slower than Platinum when it comes to Pokemon battles.

**Comment by essiw** [comment]
My mistake, I meant to say pokemon platinum.

**Comment by WilliamLCobb** [comment]
I'm pretty sure I know what's causing this. Does turning on synchronous audio make much of a difference? I usually play without sound so it's something I don't test very often

**Comment by essiw** [comment]
I just tried synchronous audio, and I am unsure if it makes a difference, because synchronous audio really sounds distorted :| So the speed might be better with synchronous audio, but I am not sure. Because of the distortion in the synchronous audio I never use that setting.

**Comment by Pepperoony** [comment]
Synchronous sound totally gets rid of the sound lag but....it some games it totally ruins the sound. Without it, most music is off tempo, and I can tell the difference with synchronous sound turned on.


--


Luke Rooney/ Director of Productions
luke.rooney@servitehs.org(mailto:luke_rooney@icloud.com)/(562)-360-4258(tel:(562)-360-4258)


Trinity Corporation
ServiteHS.org
1952 W La Palma Ave, Anaheim(x-apple-data-detectors://5/2)

**Comment by essiw** [comment]
After checking it with a less sound problematic game than pokemon platinum (I tried ace attorney: justice for all) putting synchronous audio on does remove the sound lag al to getter. Still it is horrible to listen to :| So if there is a possibility to either reduce the noise in the synchronous audio (there are some of filters that do this in desmume) that might already do the trick. Best would of course be to make the sound run more smoothly and sound like it is now, because right now it sounds perfectly, except for it being of pace.

**Comment by FrostyFoxie** [comment]
Maybe inds need to push a bit more on the ram "devices with 1GB"

**Comment by DarknesGaming** [comment]
The RAM speed for different devices vary. RAM, CPU and GPU influence the performance of devices.
Plus, some of that RAM is needed for system services so your device doesn't go haywire and shut down on you. If iNDS were to push harder on devices with 1 GB of RAM, the app may reduce battery life and crash older devices like the iPad 2 or 3. If that happens, we have to change the README.md again.

**Comment by inpuu** [comment]
I am using iPad Air2 and iPhone6S with non-jailbreak and these are the device with 2GB RAM. When I play the Professor Layton series with Synchronous Audio option, sound is terribly laggy. Without  Synchronous Audio option, there is no lag on sound but tempo gets into a mess so it is very weird. Especially the Professor Layton series have this problem than the other games. 
iOS Device: iPad Air2, iPhone6S
iOS version: 9.3.1
Jailbreak status: No
Download location: Compiled source code with Xcode
Current iNDS Version: 1.5.0. (1.5.1 version has a slow down problem)

**Comment by DarknesGaming** [comment]
The slow down problem in 1.5.1 should be fixed already. If not, we need to fix it and release 1.5.2.

**Comment by inpuu** [comment]
Slow down problem is fixed after I changed build configuration to Release :)

**Comment by DarknesGaming** [comment]
Well, it's nice to see you managed to fix that. Maybe that build configuration changed more than the icon when you did that. It probably turned off experimental features that may or may not cause more crashes.

**Comment by FrostyFoxie** [comment]
So version 1.5.1 will come on soon?

**Comment by DarknesGaming** [comment]
I think version 1.5.1 already exists. I'm trying to tweak it a little because the "Enable Mic" option won't stop enabling itself no matter how many times I disable it. It's annoying.

**Comment by DarknesGaming** [comment]
@WilliamLCobb just sent out 4 commits. I'm assuming this is iNDS 1.5.2.
GNU Lightning GIT is probably finished now. If not, then I must really suck at this.

**Comment by essiw** [comment]
@WilliamLCobb It has been a while, is there any updates on this issue? I also notice that is a slight decrease in framerate while I tested this, suddenly pokemon platinum runs on 56 FPS instead of the 60 it reached before...

**Comment by DarknesGaming** [comment]
I'm going to assume you got the latest version (1.6.0). This issue might be because a filter is on (xBRZ2 is on by default). Turn it off and see if it makes any difference.

**Comment by essiw** [comment]
^ nope and nope, 1.5.1.2 and I already turned that option off.

**Comment by DarknesGaming** [comment]
Try updating to iNDS 1.6.0 when it comes out to BuildStore.

**Comment by pmp174** [comment]
Idk what is up but I am running INDS on my iPad air 2 and it isn't reaching 60 fps. it only goes up to 30.

**Comment by IcYGodZ** [comment]
That is....horrible. Your device is faster than my ipad mini and it lags? wew

**Comment by DarknesGaming** [comment]
That's weird. My iPad mini 2 and iPhone 6 can run iNDS at full speed, but your iPad Air 2 can only run at half speed? That is terrible.

**Comment by FrogKnight** [comment]
I've pulled the latest code and still have sound lag on my iPhone 6 (not jailbroken). I'm wondering the lag problem is a bug that can be fixed or a performance problem that can not be solved.

**Comment by DarknesGaming** [comment]
The sound lag is still in the works. It has been mostly eliminated, but still present on resource-demanding ROMs.
